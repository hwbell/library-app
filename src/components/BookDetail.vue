<template>
  <div class="book-detail row">
    <!-- <p class="text">{{ this.book.volumeInfo.title}}</p> -->
    <!-- <i class="fas fa-times float-right" @click="closeDetail" @blur="closeDetail"></i> -->

    <i class="fas fa-times" @click="closeDetail"></i>

    <!-- image with info below -->
    <div class="image-holder col-sm-4">
      <img class="image" :src="info.imageLinks.thumbnail" alt>

      <p class="notes">{{info.publisher + `  ${info.publishedDate.slice(0, 4)}`}}</p>
      <p class="notes">{{ `${info.pageCount} pages` }}</p>

      <star-rating
        class="stars"
        :star-size="20"
        :fixed-points="2"
        :read-only="true"
        :rating="info.averageRating"
      ></star-rating>
      <p class="notes">{{ `avg of ${info.ratingsCount || 0} ratings` }}</p>
    </div>

    <!-- description and button to add / delete item 
    whether or not you see an add or delete option depends on
    the 'type' prop. if its a search detail, we can add the book.
    if its NOT a search, we can only delete it - its already on the list-->
    <div class="col-sm-8">
      <h4 class="title">{{ info.title}}</h4>
      <h5 class="authors">{{ info.authors.join(', ')}}</h5>

      <p class="text">{{info.description.slice(0,350)}}</p>

      <!-- popup -->
      <div>
        <transition name="fade">
          <div v-if="this.showPopup" class="popup">
            <i class="fas fa-check"></i>
            <p v-if="type == 'search'" class="alert">book added!</p>
            <p v-if="type !== 'search'" class="alert">book removed</p>
          </div>
        </transition>
      </div>
      <!--  -->

      <!-- add or remove option -->
      <div v-if="type == 'search'" class="add-item row">
        <i class="fas fa-plus-circle" @click="addBookToList"></i>
        <p class="add">add to list</p>
      </div>

      <div v-if="type !== 'search'" class="add-item row">
        <i class="fas fa-times-circle" @click="removeBookFromList"></i>
        <p class="add">remove from list</p>
      </div>
      <hr>
    </div>
  </div>
</template>

<script>
// use this npm package for the stars, very handy
import StarRating from "vue-star-rating";
import { setTimeout } from "timers";

export default {
  name: "book-detail",
  components: {
    StarRating
  },
  props: {
    book: {
      type: Object,
      required: true
    },
    type: {
      type: String,
      required: true
    }
  },
  data() {
    return {
      info: this.book.volumeInfo,
      showPopup: false
    };
  },
  directives: {
    focus: {
      inserted: function(el) {
        el.focus();
      }
    }
  },
  computed: {
    //
  },
  methods: {
    // just to close and exit
    closeDetail() {
      console.log("emitting closeDetail");
      this.$emit("closeDetail");
    },
    closePopup() {
      this.showPopup = false;
    },
    // to add the book to list in Search component
    addBookToList() {
      console.log("emitting addBookToList");
      this.showPopup = true;
      this.$emit("addBookToList");

      setTimeout(() => {
        this.showPopup = false;
        this.$emit("closeDetail");
      }, 1400);
    },
    // to remove the book from list in Search component
    removeBookFromList() {
      console.log("emitting removeBookFromList");
      this.showPopup = true;
      this.$emit("removeBookFromList");

      setTimeout(() => {
        this.showPopup = false;
        this.$emit("closeDetail");
      }, 1400);
    },
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style lang="scss" scoped>
$link-blue: rgb(0, 119, 255);

.book-detail {
  position: fixed;
  height: 60%;
  width: 70%;
  max-width: 500px;
  bottom: 0%;
  right: 15%;
  background-color: rgba(245, 245, 245, 0.98);
  box-shadow: 0px 0px 5px 2px rgba(5, 1, 71, 0.52);
  border-radius: 10px;
  overflow-y: scroll;
}
.popup {
  width: 60%;
  background-color: rgba(245, 245, 245, 0.98);
  box-shadow: 0px 0px 5px 2px rgba(5, 1, 71, 0.52);
  border-radius: 10px;
}
.image-holder {
  padding: 15px;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
}
.image {
  max-height: 230px;
  max-width: 150px;
}
.notes {
  text-align: left;
  font-size: 14px;
  padding-top: 10px;
}
.title {
  width: 80%;
  text-align: left;
  padding-top: 8px;
  font-weight: 600;
}
.authors {
  text-align: left;
  margin: 5px;
}
.text {
  text-align: left;
  font-size: 14px;
  // color: black;
}
.add-item {
  margin: 10px 0px;
}

i {
  cursor: pointer;
  transition: all 0.3s ease;
  margin: 5px;
  font-size: 20px;
}
.fa-plus-circle {
  color: $link-blue;
  &:hover {
    color: rgb(89, 0, 255);
  }
}
.fa-times-circle {
  color: rgba(255, 0, 76, 0.726);
  &:hover {
    color: rgb(255, 0, 0);
  }
}
.fa-times {
  z-index: 2;
  position: absolute;
  top: 5px;
  right: 5px;
  &:hover {
    color: grey;
  }
}
.alert {
  font-size: 12px;
}

// make the scrollbar a little softer
/* width */
::-webkit-scrollbar {
  width: 10px;
}

/* Track */
::-webkit-scrollbar-track {
  background: rgb(255, 255, 255);
  border-radius: 6px;
  margin: 4px;
}

/* Handle */
::-webkit-scrollbar-thumb {
  background: rgba(180, 180, 180, 0.95);
  border-radius: 6px;
}

/* Handle on hover */
// ::-webkit-scrollbar-thumb:hover {
//   background: rgba(5, 120, 250, 0.95);
// }
</style>