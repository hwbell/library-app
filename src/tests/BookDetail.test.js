// Import Vue and the component being tested
import { shallowMount, mount } from '@vue/test-utils'

import Vue from 'vue'
import BookDetail from '../components/BookDetail.vue';

// this components needs a lot of properties. this is a sample obj from the api
const mockData = {
  "kind": "books#volume",
  "id": "antycYtRXOUC",
  "etag": "1R+uvzDKaFw",
  "selfLink": "https://www.googleapis.com/books/v1/volumes/antycYtRXOUC",
  "volumeInfo": {
    "title": "California",
    "subtitle": "A History",
    "authors": [
      "Kevin Starr"
    ],
    "publisher": "Random House Digital, Inc.",
    "publishedDate": "2007",
    "description": "A definitive, single-volume history of the Golden State ranges from the earliest Native American cultures, through the Spanish and Mexican eras, the Gold Rush, and rise of Hollywood, to the twenty-first century, chronicling the events, places, and personalities that have shaped California. Reprint. 15,000 first printing.",
    "industryIdentifiers": [
      {
        "type": "ISBN_13",
        "identifier": "9780812977530"
      },
      {
        "type": "ISBN_10",
        "identifier": "081297753X"
      }
    ],
    "readingModes": {
      "text": false,
      "image": false
    },
    "pageCount": 370,
    "printType": "BOOK",
    "categories": [
      "History"
    ],
    "averageRating": 2.5,
    "ratingsCount": 4,
    "maturityRating": "NOT_MATURE",
    "allowAnonLogging": false,
    "contentVersion": "0.2.2.0.preview.0",
    "panelizationSummary": {
      "containsEpubBubbles": false,
      "containsImageBubbles": false
    },
    "imageLinks": {
      "smallThumbnail": "http://books.google.com/books/content?id=antycYtRXOUC&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api",
      "thumbnail": "http://books.google.com/books/content?id=antycYtRXOUC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
    },
    "language": "en",
    "previewLink": "http://books.google.com/books?id=antycYtRXOUC&printsec=frontcover&dq=history&hl=&cd=1&source=gbs_api",
    "infoLink": "http://books.google.com/books?id=antycYtRXOUC&dq=history&hl=&source=gbs_api",
    "canonicalVolumeLink": "https://books.google.com/books/about/California.html?hl=&id=antycYtRXOUC"
  },
  "saleInfo": {
    "country": "US",
    "saleability": "NOT_FOR_SALE",
    "isEbook": false
  },
  "accessInfo": {
    "country": "US",
    "viewability": "PARTIAL",
    "embeddable": true,
    "publicDomain": false,
    "textToSpeechPermission": "ALLOWED_FOR_ACCESSIBILITY",
    "epub": {
      "isAvailable": false
    },
    "pdf": {
      "isAvailable": false
    },
    "webReaderLink": "http://play.google.com/books/reader?id=antycYtRXOUC&hl=&printsec=frontcover&source=gbs_api",
    "accessViewStatus": "SAMPLE",
    "quoteSharingAllowed": false
  },
  "searchInfo": {
    "textSnippet": "A definitive, single-volume history of the Golden State ranges from the earliest Native American cultures, through the Spanish and Mexican eras, the Gold Rush, and rise of Hollywood, to the twenty-first century, chronicling the events, ..."
  }
}
describe('v', () => {

  let someProps = {
    propsData: {
      book: mockData,
      type: 'search'
    }
  }

  it('renders correctly', () => {
    const wrapper = mount(BookDetail,someProps)

    expect(wrapper.element).toMatchSnapshot()
  })

  const wrapper = shallowMount(BookDetail, someProps)
  it('is a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  })
  // it('contains navigator selector', () => {
  //   expect(wrapper.html()).toContain('id="navigator"')
  // })
  // it('contains links-holder selector', () => {
  //   expect(wrapper.html()).toContain('class="links-holder"')
  // })
  

})

