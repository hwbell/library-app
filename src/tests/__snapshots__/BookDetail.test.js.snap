// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v renders correctly 1`] = `
<div
  class="book-detail row"
>
  <i
    class="fas fa-times"
  />
   
  <div
    class="image-holder col-sm-4"
  >
    <img
      alt=""
      class="image"
      src="http://books.google.com/books/content?id=antycYtRXOUC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
    />
     
    <p
      class="notes"
    >
      Random House Digital, Inc.  2007
    </p>
     
    <p
      class="notes"
    >
      370 pages
    </p>
     
    <div
      class="stars vue-star-rating"
      data-v-34cbeed1=""
    >
      <div
        class="vue-star-rating"
        data-v-34cbeed1=""
      >
        <span
          class="vue-star-rating-star"
          data-v-34cbeed1=""
          style="margin-right: 0px;"
        >
          <svg
            class="vue-star-rating-star"
            data-v-21f5376e=""
            data-v-34cbeed1=""
            height="20"
            step="100"
            viewBox="0 0 20 20"
            width="20"
          >
            <lineargradient
              data-v-21f5376e=""
              id="164k3f"
              x1="0"
              x2="100%"
              y1="0"
              y2="0"
            >
              <stop
                data-v-21f5376e=""
                offset="100%"
                stop-color="#ffd055"
              />
               
              <stop
                data-v-21f5376e=""
                offset="100%"
                stop-color="#d8d8d8"
              />
            </lineargradient>
             
            <filter
              data-v-21f5376e=""
              filterUnits="userSpaceOnUse"
              height="130%"
              id="gcmoqg"
              width="130%"
            >
              <fegaussianblur
                data-v-21f5376e=""
                result="coloredBlur"
                stdDeviation="0"
              />
               
              <femerge
                data-v-21f5376e=""
              >
                <femergenode
                  data-v-21f5376e=""
                  in="coloredBlur"
                />
                 
                <femergenode
                  data-v-21f5376e=""
                  in="SourceGraphic"
                />
              </femerge>
            </filter>
             
            <polygon
              data-v-21f5376e=""
              fill="url(#164k3f)"
              filter="url(#gcmoqg)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#fff"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#164k3f)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#999"
              stroke-linejoin="miter"
              stroke-width="0"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#164k3f)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
            />
          </svg>
        </span>
        <span
          class="vue-star-rating-star"
          data-v-34cbeed1=""
          style="margin-right: 0px;"
        >
          <svg
            class="vue-star-rating-star"
            data-v-21f5376e=""
            data-v-34cbeed1=""
            height="20"
            step="100"
            viewBox="0 0 20 20"
            width="20"
          >
            <lineargradient
              data-v-21f5376e=""
              id="nkqy4a"
              x1="0"
              x2="100%"
              y1="0"
              y2="0"
            >
              <stop
                data-v-21f5376e=""
                offset="100%"
                stop-color="#ffd055"
              />
               
              <stop
                data-v-21f5376e=""
                offset="100%"
                stop-color="#d8d8d8"
              />
            </lineargradient>
             
            <filter
              data-v-21f5376e=""
              filterUnits="userSpaceOnUse"
              height="130%"
              id="ks3g1"
              width="130%"
            >
              <fegaussianblur
                data-v-21f5376e=""
                result="coloredBlur"
                stdDeviation="0"
              />
               
              <femerge
                data-v-21f5376e=""
              >
                <femergenode
                  data-v-21f5376e=""
                  in="coloredBlur"
                />
                 
                <femergenode
                  data-v-21f5376e=""
                  in="SourceGraphic"
                />
              </femerge>
            </filter>
             
            <polygon
              data-v-21f5376e=""
              fill="url(#nkqy4a)"
              filter="url(#ks3g1)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#fff"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#nkqy4a)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#999"
              stroke-linejoin="miter"
              stroke-width="0"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#nkqy4a)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
            />
          </svg>
        </span>
        <span
          class="vue-star-rating-star"
          data-v-34cbeed1=""
          style="margin-right: 0px;"
        >
          <svg
            class="vue-star-rating-star"
            data-v-21f5376e=""
            data-v-34cbeed1=""
            height="20"
            step="100"
            viewBox="0 0 20 20"
            width="20"
          >
            <lineargradient
              data-v-21f5376e=""
              id="9iq75f"
              x1="0"
              x2="100%"
              y1="0"
              y2="0"
            >
              <stop
                data-v-21f5376e=""
                offset="100%"
                stop-color="#ffd055"
              />
               
              <stop
                data-v-21f5376e=""
                offset="100%"
                stop-color="#d8d8d8"
              />
            </lineargradient>
             
            <filter
              data-v-21f5376e=""
              filterUnits="userSpaceOnUse"
              height="130%"
              id="3h4wap"
              width="130%"
            >
              <fegaussianblur
                data-v-21f5376e=""
                result="coloredBlur"
                stdDeviation="0"
              />
               
              <femerge
                data-v-21f5376e=""
              >
                <femergenode
                  data-v-21f5376e=""
                  in="coloredBlur"
                />
                 
                <femergenode
                  data-v-21f5376e=""
                  in="SourceGraphic"
                />
              </femerge>
            </filter>
             
            <polygon
              data-v-21f5376e=""
              fill="url(#9iq75f)"
              filter="url(#3h4wap)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#fff"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#9iq75f)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#999"
              stroke-linejoin="miter"
              stroke-width="0"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#9iq75f)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
            />
          </svg>
        </span>
        <span
          class="vue-star-rating-star"
          data-v-34cbeed1=""
          style="margin-right: 0px;"
        >
          <svg
            class="vue-star-rating-star"
            data-v-21f5376e=""
            data-v-34cbeed1=""
            height="20"
            step="100"
            viewBox="0 0 20 20"
            width="20"
          >
            <lineargradient
              data-v-21f5376e=""
              id="l8y10h"
              x1="0"
              x2="100%"
              y1="0"
              y2="0"
            >
              <stop
                data-v-21f5376e=""
                offset="0%"
                stop-color="#ffd055"
              />
               
              <stop
                data-v-21f5376e=""
                offset="0%"
                stop-color="#d8d8d8"
              />
            </lineargradient>
             
            <filter
              data-v-21f5376e=""
              filterUnits="userSpaceOnUse"
              height="130%"
              id="6zssif"
              width="130%"
            >
              <fegaussianblur
                data-v-21f5376e=""
                result="coloredBlur"
                stdDeviation="0"
              />
               
              <femerge
                data-v-21f5376e=""
              >
                <femergenode
                  data-v-21f5376e=""
                  in="coloredBlur"
                />
                 
                <femergenode
                  data-v-21f5376e=""
                  in="SourceGraphic"
                />
              </femerge>
            </filter>
             
            <polygon
              data-v-21f5376e=""
              fill="url(#l8y10h)"
              filter="url(#6zssif)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#fff"
              style="display: none;"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#l8y10h)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#999"
              stroke-linejoin="miter"
              stroke-width="0"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#l8y10h)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
            />
          </svg>
        </span>
        <span
          class="vue-star-rating-star"
          data-v-34cbeed1=""
          style="margin-right: 0px;"
        >
          <svg
            class="vue-star-rating-star"
            data-v-21f5376e=""
            data-v-34cbeed1=""
            height="20"
            step="100"
            viewBox="0 0 20 20"
            width="20"
          >
            <lineargradient
              data-v-21f5376e=""
              id="ns42a6"
              x1="0"
              x2="100%"
              y1="0"
              y2="0"
            >
              <stop
                data-v-21f5376e=""
                offset="0%"
                stop-color="#ffd055"
              />
               
              <stop
                data-v-21f5376e=""
                offset="0%"
                stop-color="#d8d8d8"
              />
            </lineargradient>
             
            <filter
              data-v-21f5376e=""
              filterUnits="userSpaceOnUse"
              height="130%"
              id="noiba"
              width="130%"
            >
              <fegaussianblur
                data-v-21f5376e=""
                result="coloredBlur"
                stdDeviation="0"
              />
               
              <femerge
                data-v-21f5376e=""
              >
                <femergenode
                  data-v-21f5376e=""
                  in="coloredBlur"
                />
                 
                <femergenode
                  data-v-21f5376e=""
                  in="SourceGraphic"
                />
              </femerge>
            </filter>
             
            <polygon
              data-v-21f5376e=""
              fill="url(#ns42a6)"
              filter="url(#noiba)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#fff"
              style="display: none;"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#ns42a6)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
              stroke="#999"
              stroke-linejoin="miter"
              stroke-width="0"
            />
             
            <polygon
              data-v-21f5376e=""
              fill="url(#ns42a6)"
              points="9.090909090909092,1.0101010101010102,3.0303030303030303,20,18.181818181818183,7.878787878787879,0,7.878787878787879,15.15151515151515,20"
            />
          </svg>
        </span>
         
        <span
          class="vue-star-rating-rating-text"
          data-v-34cbeed1=""
        >
           3.00
        </span>
      </div>
    </div>
     
    <p
      class="notes"
    >
      avg of 4 ratings
    </p>
  </div>
   
  <div
    class="col-sm-8"
  >
    <h4
      class="title"
    >
      California
    </h4>
     
    <h5
      class="authors"
    >
      Kevin Starr
    </h5>
     
    <p
      class="text"
    >
      A definitive, single-volume history of the Golden State ranges from the earliest Native American cultures, through the Spanish and Mexican eras, the Gold Rush, and rise of Hollywood, to the twenty-first century, chronicling the events, places, and personalities that have shaped California. Reprint. 15,000 first printing.
    </p>
     
    <div>
      <!---->
    </div>
     
    <div
      class="add-item row"
    >
      <i
        class="fas fa-plus-circle"
      />
       
      <p
        class="add"
      >
        add to list
      </p>
    </div>
     
    <!---->
     
    <hr />
  </div>
</div>
`;
